EE-1948 - Added JSON schemas of ArrayExperiment and ArrayAssay

M-casado opened the pull request:
Added a new folder 'schemas/' to the main directory, containing new JSON schemas for ArrayExperiment and ArrayAssay, along with the 'common' JSON schemas.

Also added a README to the new folder and some JSON datafiles that are intented to fail or pass validation based on the code on a jupyter notebook in the same folder.

See ticket: https://www.ebi.ac.uk/panda/jira/browse/EE-1948

Dona094 reviewed (commented):
README File- Apart from minor typos in the overview section everything looks good to me.
Just a small suggestion if the "Planning to modify the schemas section" could be simplified a bit say by explaining further with an example link of how the cross-file references are achieved as it can be slightly confusing for the first time

Though I  don't have much experience with JSON formats, I went through the JSON schemas, JSON datafiles, as well as the jupyter notebook, and everything looks okay to me, but I leave it to the experts of our team regarding these format reviews.

Thanks Marcos for your great work on this.

M-casado commented:
Hi @Dona094:
> README File- Apart from minor typos in the overview section everything looks good to me.

I have fixed 2-3 typos, but since you didn't mention exactly which ones you were referring to... I don't know if there are more that I didn't notice. 
> Just a small suggestion if the "Planning to modify the schemas section" could be simplified a bit say by explaining further with an example link of how the cross-file references are achieved as it can be slightly confusing for the first time

Added a bit more text explaining it and also added examples. 

Let me know if this suffices 😄

Dona094 commented:
> Hi @Dona094:
> 
> > README File- Apart from minor typos in the overview section everything looks good to me.
> 
> I have fixed 2-3 typos, but since you didn't mention exactly which ones you were referring to... I don't know if there are more that I didn't notice.
> 
> > Just a small suggestion if the "Planning to modify the schemas section" could be simplified a bit say by explaining further with an example link of how the cross-file references are achieved as it can be slightly confusing for the first time
> 
> Added a bit more text explaining it and also added examples.
> 
> Let me know if this suffices 😄

Sorry, my bad I did not mention the specific typos. Checking now, just seeing one that is differenciate under Shared definitions which should be differentiate I think. Many thanks for the detailed section now on the Planning to modify the schemas. Really comprehensive 😊

afoix commented on schemas/EGA.ArrayExperiment.json:
```diff
@@ -0,0 +1,109 @@
+      "ArrayExperiment_relationships": {
+        "type": "array",
+        "title": "ArrayExperiment relationships",
+        "description": "Comprises metadata (e.g. Source or Target) of a directional association between two entities. This relationshipos node contains all the possible relationships between metadata objects, both outside of (e.g. an Array Design Format that was submitted to ArrayExpress being linked to their microarray data within EGA) and within (e.g. an ArrayExperiment being linked to a Sample ...
```
Typo: it's written `relationshipos` and I guess the word is `relationships`

afoix replied:
You write: `we can add as many as we want` I rather use something like: `EGA`, `the user` other than a pronoun

M-casado replied:
Well spotted, in fact there is another typo in the same line. When I wrote the schemas the typo-checker was disabled, so I expect some other typos to appear. 

Regarding the pronoun used, I will also address it in my next commit.

ainacrg replied:
also the "defyning the relatiomship" > "defining the relationship" part

afoix reviewed (commented):
Minor changes/question - I've also tested `validation-notebook.ipynb` on my local instance and works OK

M-casado commented on schemas/EGA.ArrayAssay.json:
```diff
@@ -0,0 +1,89 @@
+        "type": "string",
+        "title": "Date of the ArrayAssay",
+        "description": "ISO date (format YYYY-MM-DD - e.g. '2021-05-15') when the ArrayAssay took place. If the protocols are long enough, the date shall be the day the ArrayAssay concluded.",
+        "$ref": "https://github.com/EbiEga/ega-metadata-schema/tree/main/schemas/EGA.common-definitions.json#/definitions/EGA-ISO-date-YYYY-MM-DD-pattern"
```
The ``$ref`` fields are pointers of other ``$id`` fields, whether they are in the same file or in another file. As far as I read, common practices regarding JSON schemas state that ``$id`` fields need to be identifiers, but not necessarily network-resolvable. I chose to use as the base of my identifiers the EGA-metadata GitHub URI, just so that nothing else on the web uses a similar identifier. 

See a related piece of documentation from the [Schemas Identification](https://json-schema.org/understanding-json-schema/structuring.html#schema-identification):

> Even though schemas are identified by URIs, those identifiers are not necessarily network-addressable. They are just identifiers.

ainacrg commented on schemas/EGA.ArrayExperiment.json:
```diff
@@ -0,0 +1,109 @@
+      "ArrayExperiment_protocols": {
+        "type": "array",
+        "title": "Protocols related to an ArrayExperiment",
+        "description": "Comprises metadata (e.g. Type of protocol) of a plan specification related to an ArrayExperiment, with sufficient level of detail and quantitative information to communicate it (and thus reproduce it) between investigation agents. #! Using an empty item list but definying the protocols as possible additionalItems we create the correct constraint: anything but a protocol is ...
```
Typo:
Using an empty item list but **definying** the protocols > Using an empty item list but **defining** the protocols

ainacrg commented on schemas/EGA.ArrayAssay.json:
```diff
@@ -0,0 +1,89 @@
+      "assay_files": {
+        "type": "array",
+        "title": "Assay files object from an ArrayAssay",
+        "description": "This array object contains the specific files derived from performing an hybridization of the assayed molecule to a physical array. Given the amount of technlogies available for such purpose, the allowed filetypes shall be agred upon and updated accordingly. #! Using an empty item list but definying the file object as possible additionalItems we create the correct constrai ...
```
Same line "definying" > "defining"

ainacrg replied:
This is for me to understand... assay files would be the file type, the filename, or both?

M-casado replied:
> This is for me to understand... assay files would be the file type, the filename, or both?

The ``assay_files`` node contains the following reference:
```` json
"additionalItems": { "$ref": "https://github.com/EbiEga/ega-metadata-schema/tree/main/schemas/EGA.common-definitions.json#/definitions/file_object" }
````
Which basically means that it inherits all the elements you mentioned: the file type, filename, etc. from the common-json's object ``file_object``.

ainacrg commented on schemas/EGA.ArrayAssay.json:
```diff
@@ -0,0 +1,89 @@
+      "sample_Labels": {
+        "type": "array",
+        "title": "Array of Sample-label pairs of the ArrayAssay",
+        "description": "Sample-Label pairs (format being 'EGAN[0-9]{11}:Label' - e.g. 'EGAN00000000001:Cy3') to know which samples used in this assay are labeled by which chemicals. Could be omitted if the array is of one single label and colour. #! Using an empty item list but definying the sample-label pairs as possible additionalItems we create the correct constraint: anything but a sample-lab ...
```
"definying" > "defining"

ainacrg commented on schemas/EGA.ArrayAssay.json:
```diff
@@ -0,0 +1,89 @@
+      "arrayAssay_attributes": {
+        "type": "array",
+        "title": "ArrayAssay custom attributes",
+        "description": "Custom attributes of an ArrayAssay: reusable attributes to encode tag-value pairs (e.g. Tag being 'Targeted loci' and its Value '5:63256183-63258334') with optional units (e.g. 'base pairs'). Its properties are inherited from the common-definitions.json schema. #! Using an empty item list but definying the custom attributes as possible additionalItems we create the correct ...
```
definying > defining

ainacrg commented on schemas/EGA.ArrayExperiment.json:
```diff
@@ -0,0 +1,109 @@
+      "ArrayExperiment_attributes": {
+        "type": "array",
+        "title": "ArrayExperiment custom attributes",
+        "description": "Custom attributes of an ArrayExperiment: reusable attributes to encode tag-value pairs (e.g. Tag being 'Targeted loci' and its Value '5:63256183-63258334') with optional units (e.g. 'base pairs'). Its properties are inherited from the common-definitions.json schema. #! Using an empty item list but definying the custom attributes as possible additionalItems we create the co ...
```
definying > defining

ainacrg commented on schemas/EGA.ArrayExperiment.json:
```diff
@@ -0,0 +1,109 @@
+      "ADF_files": {
+        "type": "array",
+        "title": "Array Design Format (ADF) [NCIT:C172213] file block",
+        "description": "The array design format (ADF) [NCIT:C172213] is the unique set of probes (with their coordinates) found on the microarray chip. They can be standard (sold by a company) or custom. Its format is of a spreadsheet-like tab-delimited text file with metadata header rows, followed by a multi-column table of probe information. This object shall only be used if the EGA is the one  ...
```
definying > defining

ainacrg commented on schemas/validation_tests/ArrayAssay-invalid-2.json:
```diff
@@ -0,0 +1,66 @@
+    "assay_files": [
+        {
+            "filename": "myfile1.cel",
+            "filetype": "CEL [EFO:0005630]",
```
Here I have the answer of one my comments above. Assay files will have both, filename and filetype. Are we going to use a controlled vocabulary for file types, or are we going to accept any filetype (as it is now). Just curious :)

M-casado replied:
If you take a look at ``EGA.common-definitions.json`` > ``definitions`` > ``file_object`` > ``filetype`` you will see that, at least for now, the approach is to only allow certain filetypes (those within the ``enum`` clause):

```` json
            "enum": ["CEL [EFO:0005630]", "TSV [NCIT:C164049]", "ADF [NCIT:C172213]", "FASTQ [EFO:0004155]", "FASTA [NCIT:C47845]", "SDRF [NCIT:C172211]", "IDF [NCIT:C172212]", "VCF [NCIT:C172216]", "SRA [format:3698]", "SRF [EFO:0004154]", "SFF [EFO:0004156]", "BAM [EFO:0004157]", "CRAM [format:3462]", "XLSX [format:3620]", "CSV [format:3752]"]
````

ainacrg reviewed (commented):
great job Marcos! 🥇 

I just spotted a few typos and have a couple of questions (out of curiosity).

M-casado commented:
Thanks for your review, @ainacrg! 

As has been mentioned before I expect those typos to appear a lot, since I copy-pasted many parts of the text and the typo-checker was disabled 😓 
I'll answer your questions individually in a minute and group all feedback to correct the schemas in new commits.

M-casado commented on schemas/EGA.ArrayAssay.json:
```diff
@@ -0,0 +1,89 @@
+          "type": "string",
+          "title": "Sample-label pair",
+          "pattern": "^EGAN\\d{11}:[a-zA-Z0-9 ]+$"
+        }
```
That is a good question, and for now what I chose was to make ``sample_Labels`` optional (not required), allowing only the pattern ``EGAN...:LABEL``. We could indeed make use of the aliases, but in that case we would also need:

- Providing the additional term "**centername**" of samples, unless the centername of the assay was the same as the centername that submitted the samples. 
- Additional **validation** steps upon submission checking that the aliases exist within the given centername.

Without this I feel it's a bit risky allowing aliases in those fields. Nevertheless, it is true that we may want to allow them without validating in case we simply want the user to know that the given "alias" was labeled that way. 

It's a trade off between how easy it is to provide wrong metadata and how difficult we make the process for the user.

sdelatorrep replied:
But the additional validation step will be necessary anyway, won't it? Before creating the final objects, we will need to verify that the referenced sample accessions exist. Or did I misunderstand you?

M-casado replied:
> But the additional validation step will be necessary anyway, won't it? Before creating the final objects, we will need to verify that the referenced sample accessions exist. Or did I misunderstand you?

There should be, but from the perspective of the schemas alone, that validation cannot take place. I simply thought that using EGAN IDs would be less prone to human mistakes than using centernames and aliases. Both should be eventually checked, but since that is out of the scope of these schemas...

Anyway, I can change the label nodes to use Aliases and Centernames, and just make sure that we validate them eventually. This way submissions could be done in one go, without having to wait for EGANs to be assigned to your samples.

sdelatorrep commented on schemas/EGA.ArrayAssay.json:
```diff
@@ -0,0 +1,89 @@
+        "additionalItems": {
+          "type": "string",
+          "title": "Sample-label pair",
+          "pattern": "^EGAN\\d{11}:[a-zA-Z0-9 ]+$"
```
As in the case of the ontologies, please, consider using 2 fields instead of concatenating the sample ID and the label with a colon. I believe this is clearer and easier to extend (to accept an `alias` and a `center_name`):
```json
{
   "sample_Labels": [ {
       "id": "EGAN00000000001", 
       "label": "Cy3"
   }, {
       "id": "EGAN00000000002",
       "label": "Biotin"
   }, {
       "alias": "my_sample_01",
       "center_name": "EBI-TEST",
       "label": "XXXXXX"
   }]
}
```

sdelatorrep reviewed (changes requested):
Very thorough work, Marcos, awesome! I only have a couple of suggestions :) 
EDIT: mm I can't see one comment I did. To me, it's very weird to mix snake_case and camelCase, I would stick to one or the other, but not use both in the same field, e.g. `ArrayAssay_relationships` vs `array_assay_relationships` or `arrayAssayRelationships`.

M-casado commented on schemas/EGA.ArrayAssay.json:
```diff
@@ -0,0 +1,89 @@
+        "description": "Number of samples included in the Assay (i.e. pooled into one single microarray, labelled differently). One sample corresponds to one biological replicate [EFO:0002091] (it could be the genetic content from a single cell, a tissue… from a single individual or from several individuals). Shall not be mistaken for technical replicates [CHEBI:24432] being used several times (s ...
+      },
+
+      "sample_Labels": {
```
You mean the ``sample_Label`` uppercase L, right? 

Changed to lowercase 👍

M-casado commented on schemas/EGA.common-definitions.json:
```diff
@@ -0,0 +1,531 @@
+          },
+          "checksum_method": {
+            "type": "string",
+            "title": "Checksum method [ChecksumAlgorithm]",
```
Only the checksum method, or all other nodes in which I joined the strings and ontologized terms? 

For instance, ``filetypes``, instead of:
```` json
"filetype": {
            "type": "string",
            "title": "Filetype [NCIT:C172272]",
            "description": "The nature of the content stored in an electronic file, chosen from a list of CV (e.g. CEL [EFO:0005630] or TSV [NCIT:C164049]).",
            "enum": ["CEL [EFO:0005630]", "TSV [NCIT:C164049]", "ADF [NCIT:C172213]", "FASTQ [EFO:0004155]", "FASTA [NCIT:C47845]", "SDRF [NCIT:C172211]", "IDF [NCIT:C172212]", "VCF [NCIT:C172216]", "SRA [format:3698]", "SRF [EFO:0004154]", "SFF [EFO:0004156]", "BAM [EFO:0004157]", "CRAM [format:3462]", "XLSX [format:3620]", "CSV [format:3752]", "BED [format:3003]", "IDAT [format:3578 ]", "MAP [format:3285]", "PED [format:3286]", "BIM", "FAM"]
          },
````
Something like:
```` json
"filetype": {
            "type": "string",
            "title": "Filetype [NCIT:C172272]",
            "properties": {
              "filetype_name": {
                "enum": ["CEL", "TSV", "ADF", "FASTQ", "FASTA", "SDRF", "IDF", "VCF", "SRA", "SRF", "SFF", "BAM", "CRAM", "XLSX", "CSV", "BED", "IDAT", "MAP", "PED", "BIM", "FAM"]
              },
              "filetype_CURIE": {
                "enum": ["EFO:0005630", "NCIT:C164049", "NCIT:C172213", "EFO:0004155", "NCIT:C47845", "NCIT:C172211", "NCIT:C172212", "NCIT:C172216", "format:3698", "EFO:0004154", "EFO:0004156", "EFO:0004157", "format:3462", "format:3620", "format:3752", "format:3003", "format:3578 ", "format:3285", "format:3286"]
              }
            }            
          }
````

sdelatorrep commented on schemas/EGA.ArrayAssay.json:
```diff
@@ -44,16 +44,37 @@
+          "additionalProperties": false,
+          "properties": {
+            "label": { "$ref": "https://github.com/EbiEga/ega-metadata-schema/tree/main/schemas/EGA.common-definitions.json#/definitions/array_label" },
+            "object_ids": {
```
Shouldn't it be singular, as it is an object and not an array?

M-casado replied:
Do you mean ``object_ids``?

sdelatorrep replied:
Yes, sorry for not being clearer :)

M-casado replied:
To be honest I don't really see it wrong if named in plural, since the name refers to its content, which is several object IDs. Even if the object _per se_ is not an array, it contains several other "object ID". But since it can be easily renamed, let me know if these options suit it better: 

- ``identifiers_object``
- ``object_id``
- ``core_id_object``
- Any other suggestion?

sdelatorrep reviewed (changes requested):
Good job! I only have one suggestion to rename a field :)
